### DISCUSSION TITLE
[3일차-지원] CORS

### DISCUSSION THREAD

**Discussion by smarfy99** [zzangCS/zzangInterview#14] [프론트 면접 질문]
CORS에러란 무엇이고, 발생 이유와 해결방법에 대해 말해보세요.

**Comment by ez1n** [comment]
**CORS**
- Cross-Origin Rsource  Sharing (교차 출처 리소스 공유)의 약자 → 교차 출처 : 다른 출처
- 서로 다른 도메인간에 자원을 공유하는 것을 의미하며 기본적으로 차단되어 있음

**CORS 에러** : 서로 다른 도메인간에 자원을 공유하려 할 때 발생하는 오류

**Origin** : 출처를 의미하며 Protocol + Host + Port를 합친 것
→ 출처가 같은 경우 CORS 에러는 발생하지 않는다.

**해결 방법**
- 서버에서 응답 헤더에 특정 헤더를 포함한다.
→ Access-Control-Allow-Origin : 특정 브라우저가 리소스에 접근 가능하도록 허용
→ Access-Control-Allow-Method : 특정 HTTP Method만 리소스에 접근 가능하도록 허용
→ Access-Control-Expose-Headers : 자바스크립트에서 헤더에 접근 가능하도록 허용

**Reply by ez1n** [reply]
- **CORS (Cross Origin Resource Sharing)**
    
> 한 도메인 또는 Origin의 웹 페이지가 다른 도메인을 가진 리소스에 액세스(도메인 간 요청)할 수 있게 하는 보안 메커니즘

⇒ 동일 출처 정책에 의해 등장 (same-origin policy)
⇒ 브라우저에서 출처 비교! (CORS 정책을 위반하는 리소스를 요청하더라도 정상적으로 응답한 뒤 브라우저에서 응답을 분석하여 동일 출처가 아니라고 판단하면 차단함)
    
```
💡 **동일 출처 정책 (same-origin policy)**

동일한 출처(프로토콜, 호스트명, 포트)의 리소스에만 접근하도록 제한하는 것
```
    
- **CORS 에러**

> 브라우저간의 데이터를 주고받는 과정에서, 출처가 서로 다른 사이트간의 API 요청을 할 때(SOP를 위반하는 경우) 공유를 설정하지 않았을 때 발생하는 오류

- **CORS 에러 해결**
1. 서버에서 Access-Control-Allow-Origin 헤더 세팅하기
2. 프론트엔드에서 http-proxy-middleware 설치하여 setupProxy 파일 생성

**Reply by ez1n** [reply]
https://patrick-f.tistory.com/7

**Comment by smarfy99** [comment]
CORS(Cross-Origin Resource Sharing)으로 서로 다른 도메인에 대한 요청을 보안상 제한하는 정책이다. 
그렇기 때문에 다른 포트번호나 도메인에 접속하려 하면 CORS 에러가 발생한다.
해결방법
- 클라이언트에서 HTTP요청의 헤더에 origin을 담아 전달
- 서버는 응답헤더에 Access-Control-Allow-origin을 담아 클라이언트로 전달
- 클라이언트에서 origin과 서버에서 보내준 Access-Control-Allow-Origin을 비교

**Reply by smarfy99** [reply]
https://inpa.tistory.com/entry/WEB-%F0%9F%93%9A-CORS-%F0%9F%92%AF-%EC%A0%95%EB%A6%AC-%ED%95%B4%EA%B2%B0-%EB%B0%A9%EB%B2%95-%F0%9F%91%8F
